File: solutions/cpp/collatz-conjecture/1/collatz_conjecture.cpp
#include "collatz_conjecture.h"

#include <stdexcept>

namespace collatz_conjecture {

int steps(int n) {
  if (n < 1)
    throw std::domain_error("underflow");

  int steps;
  for (steps = 0; n > 1; steps++) {
    if ((n & 1) == 0) { // even
      n >>= 1;          // ÷2
    } else {            // odd
      n = 3 * n + 1;    // 3n + 1
    }
  }
  return steps;
}

} // namespace collatz_conjecture

File: solutions/ruby/collatz-conjecture/1/collatz_conjecture.rb
# CollatzConjecture class
class CollatzConjecture
  def self.steps(n, steps=0)
    raise ArgumentError if n < 1

    return n == 1 ? steps
           : n.even? ? self.steps(n / 2, steps + 1)
           : n.odd? ? self.steps(n * 3 + 1, steps + 1)
           : '?'
  end
end
